### PR TITLE
New Resource: aws_cloudsearch_domain

### PR DESCRIPTION
**Pull request by Mi3-14159**
Hi, this adds a new cloudsearch resource to terraform.

### CODE REVIEW

**Comment by maciej-rosiek** [comment]
nice

File: aws/resource_aws_cloudsearch_service.go [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+
+		Schema: map[string]*schema.Schema{
+			"domain_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"instance_type": {
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (Ninir) [inline aws/resource_aws_cloudsearch_service.go]: Could you point me to the documentation you used for this parameters?
Only found http://docs.aws.amazon.com/cloudsearch/latest/developerguide/API_ScalingParameters.html, but nothing being required :/

File: aws/resource_aws_cloudsearch_service.go [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+
+	return nil
+}
+
+func updateScalingParameters(d *schema.ResourceData, meta interface{}, conn *cloudsearch.CloudSearch) error {
+	input := cloudsearch.UpdateScalingParametersInput{
+		DomainName: aws.String(d.Get("domain_name").(string)),
+		ScalingParameters: &cloudsearch.ScalingParameters{
+			DesiredInstanceType:     aws.String(d.Get("instance_type").(string)),
+			DesiredPartitionCount:   aws.Int64(int64(d.Get("partition_count").(int))),
```
Comment (Ninir) [inline aws/resource_aws_cloudsearch_service.go]: As per http://docs.aws.amazon.com/cloudsearch/latest/developerguide/API_ScalingParameters.html, it is valid only if the instance is `m3.2xlarge`

> DesiredPartitionCount
> The number of partitions you want to preconfigure for your domain. Only valid when you select m3.2xlarge as the instance type.

File: aws/resource_aws_cloudsearch_service.go [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+
+	_, err := conn.DeleteIndexField(&input)
+	return err
+}
+
+var parseError = func(d string, t string) error {
+	return fmt.Errorf("can't convert default_value '%s' of type '%s' to int", d, t)
+}
+
+func getOk(index map[string]interface{}, prop string, t interface{}) error {
```
Comment (Ninir) [inline aws/resource_aws_cloudsearch_service.go]: Could we reword this method? I do not feel comfortable using this, knowing a GetOk method exists on the schema :/
It could be nice also to add a comment here, as it is not obvious about what it does at the moment :/

File: aws/resource_aws_cloudsearch_service.go [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+		// Handle Removal
+		for k := range old {
+			if _, ok := new[k]; !ok {
+				deleteIndexField(d.Get("domain_name").(string), k, conn)
+			}
+		}
+
+		for _, v := range new {
+			// Handle replaces & additions
+			err := defineIndexField(d.Get("domain_name").(string), v.(map[string]interface{}), conn)
```
Comment (Ninir) [inline aws/resource_aws_cloudsearch_service.go]: Can you move the function body here? Since we use it once... 🤷‍♂️ thougths?

File: aws/resource_aws_cloudsearch_service.go [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+		return err
+	}
+
+	err = updateAccessPolicy(d, meta, conn)
+	if err != nil {
+		return err
+	}
+
+	if updated {
+		_, err := conn.IndexDocuments(&cloudsearch.IndexDocumentsInput{
```
Comment (Ninir) [inline aws/resource_aws_cloudsearch_service.go]: Could you explain what this is used for please? not sure to understand. Thanks! 👍

Comment (Mi3-14159) [reply aws/resource_aws_cloudsearch_service.go]: Hi,
as i understood if you change the index options you have to rebuild the search index.
http://docs.aws.amazon.com/cloudsearch/latest/developerguide/indexing.html

File: aws/resource_aws_cloudsearch_service.go (hunk 3 above)
Comment (Mi3-14159) [reply aws/resource_aws_cloudsearch_service.go]: You are absolutely right.

File: aws/resource_aws_cloudsearch_service.go (hunk 4 above)
Comment (Mi3-14159) [reply aws/resource_aws_cloudsearch_service.go]: In my opinion it's ok. Now we have more functions but they all are relatively small and easy to read. If i move the body of defineIndexField into the loop i will have to move deleteIndexField too into the other loop and the function defineIndexFields will grow to a monster.

Have I convinced you? :)

**Comment by Mi3-14159** [comment]
Hi @Ninir,

i'm pretty fine with the state now.
So go for the merge :)

File: aws/resource_aws_cloudsearch_domain.go [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+						},
+					},
+				},
+			},
+			"access_policy": {
+				Type:             schema.TypeString,
+				ValidateFunc:     validateIAMPolicyJson,
+				Required:         true,
+				DiffSuppressFunc: suppressEquivalentAwsPolicyDiffs,
+			},
```
Comment (Ninir) [inline aws/resource_aws_cloudsearch_domain.go]: We are lacking a few attributes to retrieve when creating the CloudSearch domain:
* ARN
* DomainID
* SearchInstanceCount
* SearchInstanceType
* SearchPartitionCount
* SearchService Endpoint
* DocService Endpoint
* Limits

Also, some attributes here can be set as Computed, as values can be retrieved from the API

What do you think?

File: aws/resource_aws_cloudsearch_domain.go [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+}
+
+func resourceAwsCloudSearchDomainCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cloudsearchconn
+
+	input := cloudsearch.CreateDomainInput{
+		DomainName: aws.String(d.Get("domain_name").(string)),
+	}
+
+	output, err := conn.CreateDomain(&input)
```
Comment (Ninir) [inline aws/resource_aws_cloudsearch_domain.go]: It would be nice to add some debug here, as in:

`log.Printf("[DEBUG] Reading CloudWatch Dashboard: %#v", input)`

File: aws/resource_aws_cloudsearch_domain.go [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+	if err != nil {
+		return fmt.Errorf("%s %q", err, d.Get("domain_name").(string))
```
Comment (Ninir) [inline aws/resource_aws_cloudsearch_domain.go]: It would be really nice to expose that we are actually failing creating the resource, as in:

`return fmt.Errorf("Creating CloudSearch Domain failed: %s", err)`

File: aws/resource_aws_cloudsearch_domain.go [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+			DesiredInstanceType:     aws.String(d.Get("instance_type").(string)),
+			DesiredReplicationCount: aws.Int64(int64(d.Get("replication_count").(int))),
+		},
+	}
+
+	if d.Get("instance_type").(string) == "search.m3.2xlarge" {
+		input.ScalingParameters.DesiredPartitionCount = aws.Int64(int64(d.Get("partition_count").(int)))
+	}
+
+	_, err := conn.UpdateScalingParameters(&input)
```
Comment (Ninir) [inline aws/resource_aws_cloudsearch_domain.go]: Can you log the input options, so that we provide more debug for users?
It would be really nice to do that for each update method! thougths?

File: aws/resource_aws_cloudsearch_domain.go [hunk 10]
```diff
@@ -0,0 +1,486 @@
+package aws
```
Comment (Ninir) [inline aws/resource_aws_cloudsearch_domain.go]: We should definitely add tests for this resource, so that it's easier to manage create, update, deletes, and check everything is fine in an automated way.
You can look at other resources if needed

File: website/docs/r/cloudsearch_domain.html.markdown [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+* `name` - (Required) Represents the property field name.
+* `type` - (Required) Represents the property type. It can be one of `int,int-array,double,double-array,literal,literal-array,text,text-array,date,date-array,lation` [AWS's Docs](http://docs.aws.amazon.com/cloudsearch/latest/developerguide/configuring-index-fields.html)
+* `search` - (Required) You can set whether this index should be searchable or not. (index of type text ist always searchable)
+* `facet` - (Optional) You can get facet information by enabling this.
+* `return` - (Required) You can enable returning the value of all searchable fields.
+* `sort` - (Optional) You can enable the property to be sortable.
+* `highlight` - (Optional) You can highlight information.
+* `analysis_scheme` - (Optional) Only needed with type `text`. [AWS's Docs - supported languages](http://docs.aws.amazon.com/cloudsearch/latest/developerguide/text-processing.html)
+* `default_value` - (Optional) The default value.
```
Comment (Ninir) [inline website/docs/r/cloudsearch_domain.html.markdown]: We will also need to add attributes in the documentation, the ones we will retrieve :)

File: vendor/vendor.json [hunk 12]
```diff
@@ -341,6 +341,14 @@
...
 		},
 		{
+			"checksumSHA1": "korrMGve3O88cRKjQ+27MK4CqAM=",
+			"path": "github.com/aws/aws-sdk-go/service/cloudsearch",
+			"revision": "124a6dabcd822d9b88383d8e2186e696a12d791d",
+			"revisionTime": "2017-09-13T19:04:32Z",
+			"version": "v1.10.44",
+			"versionExact": "v1.10.44"
+		},
+		{
```
Comment (Ninir) [inline vendor/vendor.json]: Do you think you could extract all the vendor logic to another PR? would help making the review easier, and be in sync with another resources being created, along with the SDK version being regularly updated! 😄

**Comment by radeksimko** [comment]
Hi @Mi3-14159 
thanks for the PR.

Are you still interested in finishing it? Eventually are you happy for someone else to pick it up and finish it?

**Comment by Mi3-14159** [comment]
Yes i am but i don't have any spare time to finish this.
Anyone feel free to contribute.

**Comment by bflad** [comment]
Hi Everyone 👋 Since this pull request has been idle for a long time without any movement forward, we are opting to close this pull request since it will not be merged in its current state and have created a tracking issue for Amazon CloudSearch support since one was not present before: https://github.com/terraform-providers/terraform-provider-aws/issues/7833

We encourage 👍upvotes on the feature request issue to help prioritize support for the service and for anyone who may be interested in picking this back up to comment in the tracking issue and use the author's contributions here if possible. Thanks for your understanding.
